python/cql2.pyi

```diff
@@ -26,6 +26,42 @@ class Expr:
             >>> expr = Expr.from_path("fixtures/text/example01.txt")
         """
 
+    @staticmethod
```

Comment:
Shouldn't these be `classmethod` and not `staticmethod`, since they're returning a new instance of `Expr`?

Reply:
well it shouldn't be anything I think

Reply:
at least I've never used `@staticmethod` I  think

Reply:
@kylebarron I thought so too at first, but I didn't want to bother figuring out how to dynamically init a subclass of `Expr` — I felt like `staticmethod` was more appropriate because you'll always get an `Expr` back, even if you call it from a subclass?

Reply:
but a class method should be written as 
```python 
def parse_text(cls, s: str) -> "Expr"
```

Reply:
It needs a decorator because it doesn't take `self`. But I usually think of class constructors as `classmethod` not `staticmethod`.

Defining it as a class method is primarily importing when subclassing, so that the correct subclass's constructor is called. That probably isn't something we're going to hit here, but seeing `classmethod` tells me "oh this is a constructor" rather than a `staticmethod` that says "this is just an associated function of the class"

Reply:
I'm not sure to understand why they are class/staticmethod as well

Reply:
> I'm not sure to understand why they are class/staticmethod as well

It's a good point, we could just dodge by changing them to:

```python
def cql2.parse_json(s) -> Expr:
   ...
```

If we want to go free-function, we'd want to change `parse_file` as well.

Reply:
@gadomski I don't think it technically matters here, because I don't think you can subclass a rust `PyClass` unless you [explicitly allow subclassing](https://pyo3.rs/v0.22.4/class.html?highlight=subclass#customizing-the-class). So it's just defining the logical intention here.

Reply:
> @gadomski I don't think it technically matters here, because I don't think you can subclass a rust PyClass unless you [explicitly allow subclassing](https://pyo3.rs/v0.22.4/class.html?highlight=subclass#customizing-the-class). So it's just defining the logical intention here.

👍🏼  cool, it's nice to have that protection. Feels like we should just move to free functions, since all this would be confusing to a non-rust-aware Python user anyways.